perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/modules/Vsup.java

```diff
@@ -40,55 +46,98 @@
+						} catch (ExtSourceNotExistsException | UserExtSourceNotExistsException | UserNotExistsException ex) {
+							log.warn("Couldn't find or set user to application {} by RC: {}", app, ex);
+						}
+					}
+					break;
+				}
+			}
+		}
+	}
+
+	@Override
+	public Application beforeApprove(PerunSession session, Application app) throws RegistrarException, PrivilegeException {
+
+		List<ApplicationFormItemData> data = registrar.getApplicationDataById(session, app.getId());
+
+		if (app.getUser() == null) {
+
+			for (ApplicationFormItemData item : data) {
+				if (item.getFormItem() != null &&
+					Objects.equals(AttributesManager.NS_USER_ATTR_DEF + ":birthNumber", item.getFormItem().getPerunDestinationAttribute())) {
+
+					// if application contains birth number, try to map to existing user
+					String rc = item.getValue();
+					if (rc != null && !rc.isEmpty()) {
+
+						try {
+							User user = ((PerunBl) session.getPerun()).getUsersManagerBl().getUserByExtSourceNameAndExtLogin(session, "RC", rc);
+							app.setUser(user);
+							registrar.updateApplicationUser(session, app);
+							log.debug("Existing user found by RC for {}", app);
 						} catch (Exception ex) {
-							log.warn("Couldn't find or set user to application {} by RC: {}", application, ex);
+							log.warn("Couldn't find or set user to application {} by RC: {}", app, ex);
 						}
 
 						// associate existing user with the identity used on registration form
-						if (application.getUser() != null) {
+						if (app.getUser() != null) {
 							PerunBl perunBl = (PerunBl)session.getPerun();
-							ExtSource es = perunBl.getExtSourcesManager().checkOrCreateExtSource(session, application.getExtSourceName(), application.getExtSourceType());
-							UserExtSource ues = new UserExtSource(es, application.getExtSourceLoa(), application.getCreatedBy());
+							ExtSource es = perunBl.getExtSourcesManager().checkOrCreateExtSource(session, app.getExtSourceName(), app.getExtSourceType());
+							UserExtSource ues = new UserExtSource(es, app.getExtSourceLoa(), app.getCreatedBy());
 							try {
-								ues = perunBl.getUsersManagerBl().addUserExtSource(session, application.getUser(), ues);
-								log.debug("{} associated with {} from application {}", application.getUser(), ues, application);
+								ues = perunBl.getUsersManagerBl().addUserExtSource(session, app.getUser(), ues);
+								log.debug("{} associated with {} from application {}", app.getUser(), ues, app);
 							} catch (UserExtSourceExistsException ex) {
 								// we can ignore, user will be paired with application
-								log.warn("{} already had identity associated from application {}", application.getUser(), application);
+								log.warn("{} already had identity associated from application {}", app.getUser(), app);
+							}
+						} else {
+							try {
+								Member member = ((PerunBl) session.getPerun()).getMembersManagerBl().getMemberByUser(session, app.getVo(), app.getUser());
```

Comment:
Isn't here the `app.getUser()` always null? This should probably be located elsewhere.

Reply:
Yes, it should have been in the other branch of if, thank you.